lib/force.go

```diff
@@ -1101,104 +1105,71 @@ func (f *Force) _makeHttpRequestWithoutRetry(input *httpRequestInput) (*http.Res
 	if err != nil {
 		return nil, err
 	}
-
-	// In general, a 401 or 403 will be treated as SessionExpiredError,
-	// but there are some special cases. We handle InvalidSessionId in XML errors explicitly;
-	// we also handle rate limit issues, which are a 403, explicitly, so we don't keep refreshing.
-	// So we pull these out and handle them in this if/else;
-	// if we don't find them, we create a fallback error.
-	// Then if it's a 401/403, treat the session expired (discard the fallback error);
-	// Finally, return the fallback error.
+	return res, f._coerceHttpError(res, body)
+}
+
+// Custom HTTP error handling, which is not obvious or trivial because Salesforce returns
+// many shapes and forms of errors.
+//
+// The returned error is never nil.
+//
+// In general, a 401 or 403 will be treated as SessionExpiredError,
+// but there are some special cases. We handle InvalidSessionId in XML errors explicitly;
+// we also handle rate limit issues, which are a 403, explicitly, so we don't keep refreshing.
+// So we pull these out and handle them in this if/else;
+// if we don't find them, we create a fallback error based on the actual response shape.
+// Then if it's a 401/403, treat the session expired (discard the fallback error);
+// Finally, return the fallback error.
+func (f *Force) _coerceHttpError(res *http.Response, body []byte) error {
 	var fallbackErr error
-	if strings.HasPrefix(res.Header.Get("Content-Type"), "application/xml") {
+	if strings.HasPrefix(res.Header.Get("Content-Type"), string(ContentTypeXml)) {
 		var fault LoginFault
-		xml.Unmarshal(body, &fault)
+		if err := internal.XmlUnmarshal(body, &fault); err != nil {
+			return err
+		}
 		if fault.ExceptionCode == "InvalidSessionId" {
-			return nil, SessionExpiredError
+			return SessionExpiredError
+		} else if fault.ExceptionCode != "" {
+			return fault
 		}
 	} else {
-		var messages []ForceError
-		json.Unmarshal(body, &messages)
-		if len(messages) > 0 && messages[0].ErrorCode == "REQUEST_LIMIT_EXCEEDED" {
-			return nil, APILimitExceededError
-		} else if len(messages) > 0 && messages[0].ErrorCode == "API_CURRENTLY_DISABLED" {
-			return nil, APIDisabledForUser
-		} else if len(messages) > 0 {
-			fallbackErr = errors.New(messages[0].Message)
-		} else {
-			fallbackErr = errors.New(string(body))
+		var errors ForceErrors
+		if err := internal.JsonUnmarshal(body, &errors); err != nil {
+			return err
 		}
-	}
-
-	if res.StatusCode == 401 || res.StatusCode == 403 {
-		return nil, SessionExpiredError
-	}
-	return nil, fallbackErr
-}
-
-func (f *Force) httpPostCSV(url string, data string, requestOptions ...func(*http.Request)) (body []byte, err error) {
-	body, err = f.httpPostWithContentType(url, data, "text/csv", requestOptions...)
```

Comment:
When adding a batch to a CSV query job, the Content-Type needs to be `text/csv`.  Sending `application/csv` return an error, e.g.

```
$ force bulk query -w Account "SELECT Id FROM Account"
ERROR: InvalidBatch: Wrong content-type for batch (application/csv), job is of type: text/csv
```

Reply:
Whoops, that was added a while ago. Fixed.